pkg/env/certificateRefresher.go

```diff
@@ -0,0 +1,103 @@
+package env
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CertGetter func(context.Context, string) (*rsa.PrivateKey, []*x509.Certificate, error)
+
+type RefreshingCertificateManager interface {
+	StartPoller(context.Context) error
+	GetCertificates() (*rsa.PrivateKey, []*x509.Certificate)
+}
+
+type refreshingCertificate struct {
+	lock     *sync.RWMutex
+	certs    []*x509.Certificate
+	key      *rsa.PrivateKey
+	interval time.Duration
+	logger   *logrus.Entry
+	getter   func(context.Context) (*rsa.PrivateKey, []*x509.Certificate, error)
+	done     <-chan struct{}
+}
+
+func NewRefreshingCertificate(ctx context.Context, stop <-chan struct{}, logger *logrus.Entry, interval time.Duration, getter CertGetter, certificateName string) *refreshingCertificate {
+	r := &refreshingCertificate{
+		lock:     &sync.RWMutex{},
+		logger:   logger,
+		interval: interval,
+		getter: func(ctx context.Context) (*rsa.PrivateKey, []*x509.Certificate, error) {
+			return getter(ctx, certificateName)
+		},
+		done: stop,
+	}
+
+	return r
+}
+
+func (r *refreshingCertificate) StartPoller(ctx context.Context) error {
+	// initial pull to get the certificate start
+	err := r.fetchCertificate(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.pollCertificate(ctx, r.done)
+
+	return nil
+}
+
+// GetCertificates loads the certificate from the atomic store safe to use concurently
+func (r *refreshingCertificate) GetCertificates() (*rsa.PrivateKey, []*x509.Certificate) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	return r.key, r.certs
+}
+
+// fetchCertificate access keyvault via preset getter and download new set
+// of certificates.
+// in case of failure error is returned and old certificate is left in the
+// atomic store
+func (r *refreshingCertificate) fetchCertificate(ctx context.Context) error {
+	key, certs, err := r.getter(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.lock.Lock()
+	r.key = key
+	r.certs = certs
+	r.lock.Unlock()
```

Comment:
Do a `defer r.lock.Unlock()` just after lock. Not much difference here but safer pattern in general.